### ISSUE TITLE
KeyError: 'lengths'

### ISSUE THREAD

**Issue by pseudotensor** [dask/distributed#4314]
Trying to find an older version of distributed that is not too buggy, having trouble.

rapids 0.14 pairs with 2.17 dask/disributed.

But 2.17 hits https://github.com/dask/distributed/issues/3851

Tried 2.18 as suggested there, but 2.18 hits below even for most basic test when have 2 GPUs.

Any thoughts?  I can't go to 2.27 used by rapids 0.14 because that causes many rapids tests to fail.

```
from dask.distributed import Client
from dask_cuda import LocalCUDACluster
from dask import dataframe as dd
import xgboost as xgb
def main(client):
    dask_df = dd.read_csv('creditcard.csv')
    target = 'default payment next month'
    y = dask_df['default payment next month']
    X = dask_df[dask_df.columns.difference([target])]
    dtrain = xgb.dask.DaskDMatrix(client, X, y)
    output = xgb.dask.train(client,
                            # Use GPU training algorithm
                            {'tree_method': 'gpu_hist'},
                            dtrain,
                            num_boost_round=100,
                            evals=[(dtrain, 'train')])
    booster = output['booster']  # booster is the trained model
    history = output['history']  # A dictionary containing evaluation results
    # Save the model to file
    booster.save_model('xgboost-model')
    print('Training evaluation history:', history)

    
if __name__ == '__main__':
    # `LocalCUDACluster` is used for assigning GPU to XGBoost 
    # processes. Here `n_workers` represents the number of GPUs 
    # since we use one GPU per worker process.
    with LocalCUDACluster(n_workers=2) as cluster:
        with Client(cluster) as client:
            main(client)
            

```

```
(base) jon@mr-dl10:/data/jon/h2oai.fullcondatest3$ python dask_cudf_example.py 
distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/nanny.py", line 758, in run
    await worker
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/core.py", line 236, in _
    await self.start()
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 1085, in start
    await self._register_with_scheduler()
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jon/minicondadai/lib/python3.6/_collections_abc.py", line 744, in __iter__
    yield (key, self._mapping[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/dask_cuda/device_host_file.py", line 150, in __getitem__
    return self.host_buffer[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 78, in __getitem__
    return self.slow_to_fast(key)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 65, in slow_to_fast
    value = self.slow[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/func.py", line 38, in __getitem__
    return self.load(self.d[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 505, in deserialize_bytes
    frames = merge_frames(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/utils.py", line 60, in merge_frames
    lengths = list(header["lengths"])
KeyError: 'lengths'
```

**Comment by pseudotensor** [comment]
I'm going to try 2.19, but I'm running out of options.

**Comment by pseudotensor** [comment]
I guess general question, that I can't infer, is should I be able to use latest dask/distributed with old rapids 0.14?  Are there breaking changes so that rapids would no longer work?  Was I just horribly unlucky with rapids0.15's dask/distributed 2.27?

**Comment by quasiben** [comment]
What prevents you from going to latest releases of Dask/Distributed/RAPIDS ?

**Comment by mrocklin** [comment]
> Trying to find a version of distributed that is not too buggy, having trouble.

Hi @pseudotensor !  Thanks for your issue, I'm glad to see that RAPIDS folks have found it and are responding. 

Also, while I'm sure it wasn't intentional, comments like these can come across with a negative tone.  As I'm sure you know, Dask is maintained by a broad set of people, many of whom work for free.  Negative comments on public/free issue trackers are a common source of maintainer burnout.  Again, I'm sure that this wasn't intentional, but it's maybe something to be aware of in the future.

Cheers!

**Comment by pseudotensor** [comment]
> What prevents you from going to latest releases of Dask/Distributed/RAPIDS ?

Can't use python 3.8 yet, and building rapiids stuff from scratch is not actually easy.

**Comment by pseudotensor** [comment]
> > Trying to find a version of distributed that is not too buggy, having trouble.
> 
> Hi @pseudotensor ! Thanks for your issue, I'm glad to see that RAPIDS folks have found it and are responding.
> 
> Also, while I'm sure it wasn't intentional, comments like these can come across with a negative tone. As I'm sure you know, Dask is maintained by a broad set of people, many of whom work for free. Negative comments on public/free issue trackers are a common source of maintainer burnout. Again, I'm sure that this wasn't intentional, but it's maybe something to be aware of in the future.
> 
> Cheers!

Why would it be surprising that old versions of dask have bugs?  Every software has bugs.  I was quite explicit in my goal, to use an old version of distributed that avoided particular errors.  I gave a full repro for my case, and also explained my reasoning regarding breaking changes and wondering if I could just use latest without concerns.

**Comment by quasiben** [comment]
> Can't use python 3.8 yet, and building rapiids stuff from scratch is not actually easy.

There is broad support for Python 3.7.  Are pinned to something earlier ?

**Comment by pseudotensor** [comment]
> > Can't use python 3.8 yet, and building rapiids stuff from scratch is not actually easy.
> 
> There is broad support for Python 3.7. Are pinned to something earlier ?

Yes, sorry, I should be explicit.  We are using python 3.6.  We plan to update to 3.8 soon, but it's taking time as many things have to change (3.7 wouldn't help in that regard).

**Comment by quasiben** [comment]
To answer your question:
> should I be able to use latest dask/distributed with old rapids 0.14? 

I would not expect latest dask/distributed to work that far back as a lot of changes occurred in the serialization layers between Dask and RAPIDS.  The errors you posted a probably a result of those changes.  @jakirkham do you have any thoughts here ?

**Comment by pseudotensor** [comment]
Ok thanks.  Is it possible I could patch 2.17 to avoid the original https://github.com/dask/distributed/issues/3851 problem?  I only encounter it when data gets large than about 2GB.  I'm guessing that distributed switches to protocol 5 and that's when one hits that error.  I see some work-around in the issue https://github.com/dask/distributed/issues/3851#issuecomment-638249400, which I can try.  I'm happy to also patch things if it's clear how to just fix that buffers problem.

**Comment by pseudotensor** [comment]
Thanks @quasiben  for being super responsive and helpful!

**Comment by jakirkham** [comment]
> To answer your question:
> 
> > should I be able to use latest dask/distributed with old rapids 0.14?
> 
> I would not expect latest dask/distributed to work that far back as a lot of changes occurred in the serialization layers between Dask and RAPIDS. The errors you posted a probably a result of those changes. @jakirkham do you have any thoughts here ?

There were a lot of changes related to cuDF serialization that didn't land until RAPIDS 0.15. So an older version *may* work, but also wouldn't be surprised to see issues crop up.

Is there a reason that upgrading is not an option?

**Comment by pseudotensor** [comment]
> > To answer your question:
> > > should I be able to use latest dask/distributed with old rapids 0.14?
> > 
> > 
> > I would not expect latest dask/distributed to work that far back as a lot of changes occurred in the serialization layers between Dask and RAPIDS. The errors you posted a probably a result of those changes. @jakirkham do you have any thoughts here ?
> 
> There were a lot of changes related to cuDF serialization that didn't land until RAPIDS 0.15. So an older version _may_ work, but also wouldn't be surprised to see issues crop up.
> 
> Is there a reason that upgrading is not an option?

Ya I mentioned above I cannot upgrade yet.  I'm on python 3.6 etc. see above.

Main question is if the buffers issue in 2.17 can be patched or worked around.  Otherwise 2.17 has been quite stable with rapids 0.14

**Comment by quasiben** [comment]
I think it would be a large undertaking to patch/work around

**Comment by jakirkham** [comment]
Maybe one way around this would be to copy these two files from 0.15 into 0.14 in cuDF. I think that should work, but won't make any promises here.

https://github.com/rapidsai/cudf/blob/v0.15.0/python/cudf/cudf/core/abc.py
https://github.com/rapidsai/cudf/blob/v0.15.0/python/cudf/cudf/core/buffer.py

Would be good to understand whether bumping the Python version could be an option for you or what may block it.

**Comment by pseudotensor** [comment]
I was just trying to figure out some things, but having trouble.  Seems to revolve around the use of xgboost itself.

I had previously encountered the buffers from on 2.17 very rarely with very big data, but now I see it on fresh install of rapids 0.14 using dask/distributed 2.17 when just running:

```
from dask.distributed import Client
from dask_cuda import LocalCUDACluster
from dask import dataframe as dd
import xgboost as xgb
def main(client):
    dask_df = dd.read_csv('creditcard.csv')
    target = 'default payment next month'
    y = dask_df['default payment next month']
    X = dask_df[dask_df.columns.difference([target])]
    dtrain = xgb.dask.DaskDMatrix(client, X, y)
    output = xgb.dask.train(client,
                            {'tree_method': 'gpu_hist'},
                            dtrain,
                            num_boost_round=100,
                            evals=[(dtrain, 'train')])
    booster = output['booster']  # booster is the trained model
    history = output['history']  # A dictionary containing evaluation results
    # Save the model to file
    booster.save_model('xgboost-model')
    print('Training evaluation history:', history)
    print(xgb.__version__)

    
if __name__ == '__main__':
    with LocalCUDACluster() as cluster:
        with Client(cluster) as client:
            main(client)
```

But there are complications:

1) If I run on 1/3 systems I have, I see no such problem for exact same install of packages, as confirmed by diffing list of conda list or pkgs folder names that has packages/versions.  I can't explain this.

2) Those 2/3 systems with buffers error fail now every time even with this simple test, even though I've been using the same setup for last 4 weeks without problems.  I can't explain this.

3) If I avoid all xgboost related lines (import, dask calls, etc.) then the example passes on 3/3 systems.  E.g. one can add an X.compute() at end at least so something is done.

4) If I change from cuda cluster to normal cluster, I see no buffers issues on those 2/3 systems.

```
from dask.distributed import Client, LocalCluster
from dask import dataframe as dd
import xgboost as xgb
def main(client):
    dask_df = dd.read_csv('creditcard.csv')
    target = 'default payment next month'
    y = dask_df['default payment next month']
    X = dask_df[dask_df.columns.difference([target])]
    dtrain = xgb.dask.DaskDMatrix(client, X, y)
    output = xgb.dask.train(client,
                            {'tree_method': 'gpu_hist'},
                            dtrain,
                            num_boost_round=100,
                            evals=[(dtrain, 'train')])
    booster = output['booster']  # booster is the trained model
    history = output['history']  # A dictionary containing evaluation results
    # Save the model to file
    booster.save_model('xgboost-model')
    print('Training evaluation history:', history)
    print(xgb.__version__)

    
if __name__ == '__main__':
    with LocalCluster(n_workers=1, processes=True) as cluster:
        with Client(cluster) as client:
            main(client)
```

Here is the error:

```
base) ops@mr-dl18:~/h2oai$ python dask_cudf_example2.py
/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/node.py:155: UserWarning: Port 8787 is already in use.
Perhaps you already have a cluster running?
Hosting the HTTP server on port 33393 instead
  http_address["port"], self.http_server.port
distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/nanny.py", line 758, in run
    await worker
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/core.py", line 236, in _
    await self.start()
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 1085, in start
    await self._register_with_scheduler()
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jenkins/minicondadai/lib/python3.6/_collections_abc.py", line 744, in __iter__
    yield (key, self._mapping[key])
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/dask_cuda/device_host_file.py", line 150, in __getitem__
    return self.host_buffer[key]
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 78, in __getitem__
    return self.slow_to_fast(key)
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 65, in slow_to_fast
    value = self.slow[key]
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/zict/func.py", line 38, in __getitem__
    return self.load(self.d[key])
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 502, in deserialize_bytes
    return deserialize(header, frames)
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 302, in deserialize
    return loads(header, frames)
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 64, in pickle_loads
    return pickle.loads(x, buffers=buffers)
  File "/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/protocol/pickle.py", line 64, in loads
    return pickle.loads(x, buffers=buffers)
TypeError: 'buffers' is an invalid keyword argument for this function
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP  local=tcp://127.0.0.1:59232 remote=tcp://127.0.0.1:39489>
distributed.nanny - ERROR - Failed while trying to start worker process: 'buffers' is an invalid keyword argument for this function
distributed.utils - ERROR - addresses should be strings or tuples, got None
Traceback (most recent call last):
```

FYI the 8787 is used because "rserver" is there.  But that shouldn't matter.

**Comment by pseudotensor** [comment]
FYI same thing  happens if I setup a cluster using CLI I hit:

```
distributed.protocol.pickle - INFO - Failed to deserialize b'\x80\x0 ....
```
and more

```
x00dC\x00\x00}C\x00\x80\x82C\x00\x00\x86C\x00\x006C\x00\x00|B\x00\x00\xceB\x00\x00WC\x00\x007C\x00\x00\xb4B\x00\x00|C\x00\x00\xc0\x7f\x00\x00\xc2B\x00\x00\x98C\x00\x00pC\x00\x00'
Traceback (most recent call last):
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/pickle.py", line 64, in loads
    return pickle.loads(x, buffers=buffers)
TypeError: 'buffers' is an invalid keyword argument for this function
distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/nanny.py", line 758, in run
    await worker
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/core.py", line 236, in _
    await self.start()
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 1085, in start
    await self._register_with_scheduler()
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jon/minicondadai/lib/python3.6/_collections_abc.py", line 744, in __iter__
    yield (key, self._mapping[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/dask_cuda/device_host_file.py", line 150, in __getitem__
    return self.host_buffer[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 78, in __getitem__
    return self.slow_to_fast(key)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 65, in slow_to_fast
    value = self.slow[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/func.py", line 38, in __getitem__
    return self.load(self.d[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 502, in deserialize_bytes
    return deserialize(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 302, in deserialize
    return loads(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 64, in pickle_loads
    return pickle.loads(x, buffers=buffers)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/pickle.py", line 64, in loads
    return pickle.loads(x, buffers=buffers)
TypeError: 'buffers' is an invalid keyword argument for this function
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP  local=tcp://172.16.2.210:53658 remote=tcp://172.16.2.210:8790>
distributed.nanny - ERROR - Failed while trying to start worker process: 'buffers' is an invalid keyword argument for this function

```

when I run:

```
(base) jon@mr-dl10:/data/jon/h2oai.fullcondatest3$ cat dask_cudf_example2_cluster.py
from dask.distributed import Client
from dask_cuda import LocalCUDACluster
from dask import dataframe as dd
import xgboost as xgb
def main(client):
    dask_df = dd.read_csv('creditcard.csv')
    target = 'default payment next month'
    y = dask_df['default payment next month']
    X = dask_df[dask_df.columns.difference([target])]
    dtrain = xgb.dask.DaskDMatrix(client, X, y)
    output = xgb.dask.train(client,
                            {'tree_method': 'gpu_hist'},
                            dtrain,
                            num_boost_round=100,
                            evals=[(dtrain, 'train')])
    booster = output['booster']  # booster is the trained model
    history = output['history']  # A dictionary containing evaluation results
    # Save the model to file
    booster.save_model('xgboost-model')
    print('Training evaluation history:', history)
    print(xgb.__version__)

    
if __name__ == '__main__':
    with Client(address="172.16.2.210:8790") as client:
        main(client)
         
```

**Comment by pseudotensor** [comment]
This kind of thing worked all the time before on same system and packages and overall install but always fails now.  I'm in totally broken state w.r.t. dask, any guidance?

**Comment by jakirkham** [comment]
Yeah need to get `pickle5`.

`conda install -c conda-forge pickle5`

**Comment by pseudotensor** [comment]
That didn't help.  Get same errors.  And would have thought if needed pickle5 package one of the deps would have found during the solution building.  Also, the relevant file just does import pickle, it doesn't try to use pickle5.

It seems `/home/jenkins/minicondadai/lib/python3.6/site-packages/distributed/protocol/pickle.py` has a condition for if buffers and just always passes that in, even if using conda/python native pickle.

I've never had pickle5 installed in last 4 weeks btw.  I can't explain why things have stopped working.  But as-is, I would say distributed is passing "buffers" to the loads function even if using pickle from conda/python.

**Comment by pseudotensor** [comment]
I'll try to debug my example.

**Comment by pseudotensor** [comment]
FYI If I replace import pickle with import pickle5 as pickle in pickle.py, then I still have issues:

```
distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/nanny.py", line 758, in run
    await worker
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/core.py", line 236, in _
    await self.start()
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 1085, in start
    await self._register_with_scheduler()
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/worker.py", line 811, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/jon/minicondadai/lib/python3.6/_collections_abc.py", line 744, in __iter__
    yield (key, self._mapping[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/dask_cuda/device_host_file.py", line 150, in __getitem__
    return self.host_buffer[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 78, in __getitem__
    return self.slow_to_fast(key)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/buffer.py", line 65, in slow_to_fast
    value = self.slow[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/func.py", line 38, in __getitem__
    return self.load(self.d[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 502, in deserialize_bytes
    return deserialize(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 302, in deserialize
    return loads(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 64, in pickle_loads
    return pickle.loads(x, buffers=buffers)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/pickle.py", line 64, in loads
    return pickle.loads(x, buffers=buffers)
_pickle.UnpicklingError: pickle data was truncated
Traceback (most recent call last):
  File "/home/jon/minicondadai/lib/python3.6/multiprocessing/queues.py", line 234, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/home/jon/minicondadai/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class '_pickle.UnpicklingError'>: it's not the same object as _pickle.UnpicklingError
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP  local=tcp://127.0.0.1:57796 remote=tcp://127.0.0.1:41179>
distributed.nanny - ERROR - Failed to start worker
```

**Comment by pseudotensor** [comment]
This minimal example (no xgboost or even dask operations) still does it:

```
from dask.distributed import Client
from dask_cuda import LocalCUDACluster
def main(client):
    pass

import time
if __name__ == '__main__':
    with LocalCUDACluster() as cluster:
        time.sleep(1)
        with Client(cluster) as client:
            time.sleep(1)
            main(client)

```

**Comment by pseudotensor** [comment]
I'm guessing dask is picking up on some old state, since it's showing dataframes still during pickling.  Some old job sitting on disk that dask is picking up?

**Comment by pseudotensor** [comment]
There is a folder "storage" that has tons of old from_pandas stuff I think it's picking up.

That would explain why other system works but others don't.  

Not sure about cuda vs. non-cuda issue still.  But at least it explains the historical issue that used to all work.

I presume it's possible that dask/distributed 2.18 still offer a solution to the big data problem, but I'm still confused by I ever got into situation where "buffers" was passed.  I presume it is a bug in distributed or dask_cuda.

**Comment by pseudotensor** [comment]
I'm going to guess that both problems are due to dask_cuda.  both trying wrong pickle commands and using storage on disk leading to bad re-processing of old operations?

**Comment by pseudotensor** [comment]
@jakirkham Here is repro:

https://github.com/rapidsai/dask-cuda/issues/467

**Comment by pseudotensor** [comment]
If buffers is not dask_cuda issue, what do I need to make sure distributed works?  pickle5 did not work, and is that really a requirement that is not explicit for dask/distributed?  i.e. some package you should install if you want to do big data?  Did I not RTFM?  I real *alot* of dask docs etc. and I never noticed.

**Comment by pseudotensor** [comment]
Here is debug view of the variable state.  It shows the pickler was pickle, there just is 2 buffers.  The pickle.py pickle dumps must be doing something wrong in distributed sometimes.


![image](https://user-images.githubusercontent.com/2249614/101236441-0b6ad380-3686-11eb-8377-ee166d433a84.png)

**Comment by pseudotensor** [comment]
E.g. this function:
```

def pickle_dumps(x):
    header = {"serializer": "pickle"}
    frames = [None]
    buffer_callback = lambda f: frames.append(memoryview(f))
    frames[0] = pickle.dumps(x, buffer_callback=buffer_callback)
    return header, frames
```

seems to append frames if memoryview(f) call_back gives non-None result, which it must have.

**Comment by pseudotensor** [comment]
FYI if I replace all instances of pickle package imports (i.e. import pickle or from pickle import ...) with import pickle5 as pickle etc. then I get different error:

```
    value = self.slow[key]
  File "/home/jon/minicondadai/lib/python3.6/site-packages/zict/func.py", line 38, in __getitem__
    return self.load(self.d[key])
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 502, in deserialize_bytes
    return deserialize(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 302, in deserialize
    return loads(header, frames)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 64, in pickle_loads
    return pickle.loads(x, buffers=buffers)
  File "/home/jon/minicondadai/lib/python3.6/site-packages/distributed/protocol/pickle.py", line 64, in loads
    return pickle.loads(x, buffers=buffers)
_pickle.UnpicklingError: pickle data 
```

At least in 2.17, I don't see handling of pickle5, only native pickle that would only apply for python 3.8

**Comment by jakirkham** [comment]
Would make sure that Dask and Distributed are up-to-date. Also would make sure a recent version of Cloudpickle is installed. Finally would double check that the environment used by the Client, Workers, and Scheduler all have Cloudpickle and Pickle5. After doing all of that, it may be worth running `client.get_versions(check=True)` just to make sure everything is installed as expected.

**Comment by pseudotensor** [comment]
As I mentioned import pickle is still in 2.17 code importing old pickle, not pickle5.  Only later does distributed try import pickle5, so installing pickle5 has no effect.

```
(base) jon@pseudotensor:~/h2oai.fullcondatest$ python
Python 3.6.10 | packaged by conda-forge | (default, Apr 24 2020, 16:44:11) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from dask.distributed import Client
>>> client = Client()
>>> client.get_versions(check=True)
{'scheduler': {'host': {'python': '3.6.10.final.0', 'python-bits': 64, 'OS': 'Linux', 'OS-release': '5.4.0-56-generic', 'machine': 'x86_64', 'processor': 'x86_64', 'byteorder': 'little', 'LC_ALL': 'None', 'LANG': 'en_US.UTF-8'}, 'packages': {'python': '3.6.10.final.0', 'dask': '2.17.0', 'distributed': '2.17.0', 'msgpack': '1.0.0', 'cloudpickle': '1.6.0', 'tornado': '6.0.4', 'toolz': '0.11.1', 'numpy': '1.18.5', 'lz4': None, 'blosc': None}}, 'workers': {'tcp://127.0.0.1:34443': {'host': {'python': '3.6.10.final.0', 'python-bits': 64, 'OS': 'Linux', 'OS-release': '5.4.0-56-generic', 'machine': 'x86_64', 'processor': 'x86_64', 'byteorder': 'little', 'LC_ALL': 'None', 'LANG': 'en_US.UTF-8'}, 'packages': {'python': '3.6.10.final.0', 'dask': '2.17.0', 'distributed': '2.17.0', 'msgpack': '1.0.0', 'cloudpickle': '1.6.0', 'tornado': '6.0.4', 'toolz': '0.11.1', 'numpy': '1.18.5', 'lz4': None, 'blosc': None}}, 'tcp://127.0.0.1:35603': {'host': {'python': '3.6.10.final.0', 'python-bits': 64, 'OS': 'Linux', 'OS-release': '5.4.0-56-generic', 'machine': 'x86_64', 'processor': 'x86_64', 'byteorder': 'little', 'LC_ALL': 'None', 'LANG': 'en_US.UTF-8'}, 'packages': {'python': '3.6.10.final.0', 'dask': '2.17.0', 'distributed': '2.17.0', 'msgpack': '1.0.0', 'cloudpickle': '1.6.0', 'tornado': '6.0.4', 'toolz': '0.11.1', 'numpy': '1.18.5', 'lz4': None, 'blosc': None}}, 'tcp://127.0.0.1:40115': {'host': {'python': '3.6.10.final.0', 'python-bits': 64, 'OS': 'Linux', 'OS-release': '5.4.0-56-generic', 'machine': 'x86_64', 'processor': 'x86_64', 'byteorder': 'little', 'LC_ALL': 'None', 'LANG': 'en_US.UTF-8'}, 'packages': {'python': '3.6.10.final.0', 'dask': '2.17.0', 'distributed': '2.17.0', 'msgpack': '1.0.0', 'cloudpickle': '1.6.0', 'tornado': '6.0.4', 'toolz': '0.11.1', 'numpy': '1.18.5', 'lz4': None, 'blosc': None}}, 'tcp://127.0.0.1:42655': {'host': {'python': '3.6.10.final.0', 'python-bits': 64, 'OS': 'Linux', 'OS-release': '5.4.0-56-generic', 'machine': 'x86_64', 'processor': 'x86_64', 'byteorder': 'little', 'LC_ALL': 'None', 'LANG': 'en_US.UTF-8'}, 'packages': {'python': '3.6.10.final.0', 'dask': '2.17.0', 'distributed': '2.17.0', 'msgpack': '1.0.0', 'cloudpickle': '1.6.0', 'tornado': '6.0.4', 'toolz': '0.11.1', 'numpy': '1.18.5', 'lz4': None, 'blosc': None}}}, 'client': {'host': {'python': '3.6.10.final.0', 'python-bits': 64, 'OS': 'Linux', 'OS-release': '5.4.0-56-generic', 'machine': 'x86_64', 'processor': 'x86_64', 'byteorder': 'little', 'LC_ALL': 'None', 'LANG': 'en_US.UTF-8'}, 'packages': {'python': '3.6.10.final.0', 'dask': '2.17.0', 'distributed': '2.17.0', 'msgpack': '1.0.0', 'cloudpickle': '1.6.0', 'tornado': '6.0.4', 'toolz': '0.11.1', 'numpy': '1.18.5', 'lz4': None, 'blosc': None}}}
```

**Comment by jakirkham** [comment]
Sorry I don't follow. Thought `distributed` was being updated to a newer version. If not, I don't see a path forward here.

**Comment by pseudotensor** [comment]
> Sorry I don't follow. Thought `distributed` was being updated to a newer version. If not, I don't see a path forward here.

rapids 0.14 was released with dask/distributed 2.17, but I guess it's possible it never worked, yes.

**Comment by jakirkham** [comment]
Well it was [released with 2.15+]( https://github.com/rapidsai/integration/blob/66c962be84d1a9668f11d3e0a744e61ade8613b9/conda/recipes/release-versions.yaml#L44-L47 ). Maybe trying one of the older versions works for you?

**Comment by quasiben** [comment]
I think what we are seeing here is that it is indeed challenging to find patched version of dask/distributed working with an older version of RAPIDS.  If you want to test exactly what was available during the 0.14 release you should use the [0.14 docker image](https://hub.docker.com/r/rapidsai/rapidsai/tags?page=1&ordering=last_updated&name=0.14) .  There you'll find dask/distributed 2.20.0

I'm not exactly sure where to go from here.  Happy to find 30 minutes for us to chat so I/we can better understand what you are trying to accomplish but it's unlikely that we can help you find a patch for an older version of RAPIDS

**Comment by pseudotensor** [comment]
I'm just asking how things work.  I think the dask_cuda is using a spill-over to pickle and then dask is reading those files in, and distributed fails because dask_cuda wrote to disk files that inside have extra frames distributed interprets as buffers.  That's what I have seen.  In that case, it doesn't matter what version of distributed is used, even latest only tries pickle5 if it exists, and pickle5 is not a requirement OOTB AFAIK.  Only python 3.8 would allow distributed to use protocol5.

**Comment by pseudotensor** [comment]
> I think what we are seeing here is that it is indeed challenging to find patched version of dask/distributed working with an older version of RAPIDS. If you want to test exactly what was available during the 0.14 release you should use the [0.14 docker image](https://hub.docker.com/r/rapidsai/rapidsai/tags?page=1&ordering=last_updated&name=0.14) . There you'll find dask/distributed 2.20.0
> 
> I'm not exactly sure where to go from here. Happy to find 30 minutes for us to chat so I/we can better understand what you are trying to accomplish but it's unlikely that we can help you find a patch for an older version of RAPIDS

Thanks @quasiben for pointing out docker was using 2.20.

**Comment by pseudotensor** [comment]
I'll try to learn about the spill over myself since I understand people are busy.  I know it happens when I use bigger data, and I'll just assume you guys hope it's fixed in later rapids/dask/distributed.
